controllers/storageautoscaler/storageautoscaler_controller.go

```diff
@@ -0,0 +1,255 @@
+package storageautoscaler
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	prometheusapi "github.com/prometheus/client_golang/api"
+	prometheusv1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	prometheusconfig "github.com/prometheus/common/config"
+	"github.com/prometheus/common/model"
+	ocsv1 "github.com/red-hat-storage/ocs-operator/api/v4/v1"
+	"github.com/red-hat-storage/ocs-operator/v4/controllers/platform"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// StorageAutoscalerReconciler is the reconciler for StorageAutoscaler objects
+// nolint:revive
+type StorageAutoscalerReconciler struct {
+	client.Client
+	ctx               context.Context
+	Log               logr.Logger
+	Scheme            *runtime.Scheme
+	OperatorNamespace string
+	RequeueTime       time.Duration
+	SyncMap           sync.Map
+	Event             chan event.GenericEvent
+}
+
+// SetupWithManager sets up the reconciler with the manager
+func (r *StorageAutoscalerReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// run a go routine to scrape the metrics from prometheus periodically
+	go r.scrapeMetricsPeriodically()
+
+	storageAutoscalerReconcilerController := ctrl.NewControllerManagedBy(mgr).
+		For(&ocsv1.StorageAutoScaler{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
+		// watch for generic events to trigger the reconcile
+		WatchesRawSource(source.Channel(r.Event,
+			&handler.EnqueueRequestForObject{},
+		))
+
+	return storageAutoscalerReconcilerController.Complete(r)
+}
+
+// +kubebuilder:rbac:groups="monitoring.coreos.com",resources=*,verbs=get
+
+// Reconcile reconciles the StorageAutoscaler object
+func (r *StorageAutoscalerReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	r.ctx = ctx
+	r.Log.Info("reconciling storage autoscaler", "namespace", req.Namespace, "name", req.Name)
+
+	// print the sync map
+	r.SyncMap.Range(func(key, value interface{}) bool {
+		r.Log.Info("main function sync map", "key", key.(string), "value", value.(float64))
+		return true
+	})
+
+	// todo implement the reconcile logic
+
+	return reconcile.Result{}, nil
+}
+
+func (r *StorageAutoscalerReconciler) scrapeMetricsPeriodically() {
+	for {
+		select {
+		// context of the main reconcile loop is cancelled
+		case <-context.Background().Done():
```

Comment:
I think this is creating a new context and then waiting for it to be done, which I think will be never? Should this use `ctx.Done()`?

Reply:
ctx is the context of the `StorageAutoscalerReconciler` reconcile,
I need a context of the main function(), from where SetupWithManager() is called, 
AS it doesn't have any context I thought using `context.Background()`

Do we need anything else?

Reply:
@iamniting do you have any suggestion, if there is an operator context?

Reply:
There is. If you follow my suggestion here: https://github.com/red-hat-storage/ocs-operator/pull/3071/files#r1990710825 . You'll be able to leverage operator context.

Reply:
Thanks updated